File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    static int result = 0;

  public int solution(int[] numbers, int target) {
    int answer = 0;

    dfs(numbers, target, 0, 0);

    answer = result;

    return answer;
  }

  public void dfs(int[] numbers, int target, int depth, int sum) {
    if(depth == numbers.length) {
      if(sum == target) {
        result++;
      }
      return;
    }
    
    dfs(numbers, target, depth+1, sum - numbers[depth]);
    dfs(numbers, target, depth+1, sum + numbers[depth]);

  }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
answer = 0
def solution(numbers, target):
    dfs(numbers, 0, 0, target, len(numbers))
    return answer

def dfs(numbers, total, depth, target, length):
    global answer
    if depth == length:
        if total == target:
            answer += 1
        return
    
    dfs(numbers, total+numbers[depth], depth + 1, target, length)
    dfs(numbers, total-numbers[depth], depth + 1, target, length)